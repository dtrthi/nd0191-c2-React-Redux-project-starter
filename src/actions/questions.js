import { hideLoading, showLoading } from "react-redux-loading-bar";
import { _saveQuestion, _saveQuestionAnswer } from "../_DATA";
import { addUserAnswer, addUserQuestion } from "./users";

export const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";
export const ADD_QUESTION = "ADD_QUESTION";
export const UPDATE_VOTES = "UPDATE_VOTES";

export function receiveQuestions(questions) {
  return {
    type: RECEIVE_QUESTIONS,
    questions,
  };
}

export function addQuestion(question) {
  return {
    type: ADD_QUESTION,
    question,
  };
}

export function updateVotes(questionId, option, userId) {
  return {
    type: UPDATE_VOTES,
    questionId,
    option,
    userId,
  };
}

export function handleAddQuestion(question, cb = () => {}) {
  return (dispatch) => {
    dispatch(showLoading());
    _saveQuestion(question)
      .then((newQuestion) => {
        dispatch(addQuestion(newQuestion));
        dispatch(addUserQuestion(newQuestion.author, newQuestion.id));
        dispatch(hideLoading());
        cb();
      })
      .catch(() => {
        alert("Cannot save question");
      });
  };
}

export function handleSaveQuestionAnwser(answer) {
  return (dispatch) => {
    dispatch(showLoading());
    _saveQuestionAnswer(answer).then(() => {
      dispatch(
        addUserAnswer(answer.authedUser, { [answer.qid]: answer.answer }),
      );
      dispatch(hideLoading());
    });
  };
}
