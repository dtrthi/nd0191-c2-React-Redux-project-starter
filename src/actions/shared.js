import { hideLoading, showLoading } from "react-redux-loading-bar";
import { _getQuestions, _getUsers } from "../_DATA";
import { receiveQuestions } from "./questions";
import { receiveUsers } from "./users";

export function handleInitialData() {
  return (dispatch) => {
    dispatch(showLoading());
    return Promise.all([_getUsers(), _getQuestions()])
      .then(([users, questions]) => {
        dispatch(receiveUsers(users));
        dispatch(receiveQuestions(questions));
      })
      .then(() => dispatch(hideLoading()));
  };
}
