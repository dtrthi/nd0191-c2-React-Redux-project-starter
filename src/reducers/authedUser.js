import { LOGIN, LOGOUT } from "../actions/authedUser";

export function authedUser(state = null, action) {
  switch (action.type) {
    case LOGIN:
      return action.user;
    case LOGOUT:
      return null;
    default:
      return state;
  }
}
