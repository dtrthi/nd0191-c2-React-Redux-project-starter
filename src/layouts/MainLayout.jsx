import { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { NavLink, Outlet, useLocation, useNavigate } from "react-router-dom";
import { logout } from "../actions/authedUser";
import { handleInitialData } from "../actions/shared";

function MainLayout() {
  const navigate = useNavigate();
  const authedUser = useSelector((state) => state.authedUser);
  const dispatch = useDispatch();
  const location = useLocation();
  const handleLogout = (e) => {
    e.preventDefault();
    dispatch(logout());
  };
  useEffect(() => {
    if (!authedUser) {
      navigate(`/login?redirect=${location.pathname}`);
    }
    dispatch(handleInitialData());
  }, [navigate, authedUser, dispatch, location]);

  if (!authedUser) return null;

  return (
    <div>
      <nav className="navbar navbar-expand-lg bg-body-tertiary">
        <div className="container-fluid">
          <div className="collapse navbar-collapse" id="navbarText">
            <ul className="navbar-nav me-auto mb-2 mb-lg-0">
              <li className="nav-item">
                <NavLink className="nav-link" to="/">
                  Home
                </NavLink>
              </li>
              <li className="nav-item">
                <NavLink className="nav-link" to="/leaderboard">
                  Leaderboard
                </NavLink>
              </li>
              <li className="nav-item">
                <NavLink className="nav-link" to="/add">
                  New
                </NavLink>
              </li>
            </ul>
            <span className="d-flex">
              <span className="d-flex align-items-center justify-content-center">
                <img
                  height={30}
                  width={30}
                  src={authedUser.avatarURL ?? "/default-user.jpg"}
                  alt={authedUser.name}
                />
                {authedUser.name}
              </span>
              <button className="btn" onClick={handleLogout}>
                Log out
              </button>
            </span>
          </div>
        </div>
      </nav>
      <main>
        <Outlet />
      </main>
    </div>
  );
}

export default MainLayout;
