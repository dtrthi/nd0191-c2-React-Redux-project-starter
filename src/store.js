import { legacy_createStore } from "redux";
import middlewares from "./middlewares";
import reducers from "./reducers";

export default legacy_createStore(reducers, middlewares);
