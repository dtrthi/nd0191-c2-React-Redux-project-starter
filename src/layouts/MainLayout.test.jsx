import "@testing-library/jest-dom/vitest";
import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import { login } from "../actions/authedUser";
import store from "../store";
import MainLayout from "./MainLayout";

describe("MainLayout", () => {
  it("should match snapshot", () => {
    store.dispatch(login({ id: "id" }));
    const component = render(
      <Provider store={store}>
        <MemoryRouter>
          <MainLayout />
        </MemoryRouter>
      </Provider>,
    );
    expect(component).toMatchSnapshot();
  });

  it("should have menu items", () => {
    store.dispatch(login({ id: "id" }));
    render(
      <Provider store={store}>
        <MemoryRouter>
          <MainLayout />
        </MemoryRouter>
      </Provider>,
    );
    expect(screen.getByText("Home")).toBeInTheDocument();
    expect(screen.getByText("Leaderboard")).toBeInTheDocument();
    expect(screen.getByText("New")).toBeInTheDocument();
  });
});
