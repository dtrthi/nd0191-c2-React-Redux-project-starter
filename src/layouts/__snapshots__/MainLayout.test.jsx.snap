// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MainLayout > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <nav
          class="navbar navbar-expand-lg bg-body-tertiary"
        >
          <div
            class="container-fluid"
          >
            <div
              class="collapse navbar-collapse"
              id="navbarText"
            >
              <ul
                class="navbar-nav me-auto mb-2 mb-lg-0"
              >
                <li
                  class="nav-item"
                >
                  <a
                    aria-current="page"
                    class="nav-link active"
                    data-discover="true"
                    href="/"
                  >
                    Home
                  </a>
                </li>
                <li
                  class="nav-item"
                >
                  <a
                    class="nav-link"
                    data-discover="true"
                    href="/leaderboard"
                  >
                    Leaderboard
                  </a>
                </li>
                <li
                  class="nav-item"
                >
                  <a
                    class="nav-link"
                    data-discover="true"
                    href="/add"
                  >
                    New
                  </a>
                </li>
              </ul>
              <span
                class="d-flex"
              >
                <span
                  class="d-flex align-items-center justify-content-center"
                >
                  <img
                    height="30"
                    src="/default-user.jpg"
                    width="30"
                  />
                </span>
                <button
                  class="btn"
                >
                  Log out
                </button>
              </span>
            </div>
          </div>
        </nav>
        <main />
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <nav
        class="navbar navbar-expand-lg bg-body-tertiary"
      >
        <div
          class="container-fluid"
        >
          <div
            class="collapse navbar-collapse"
            id="navbarText"
          >
            <ul
              class="navbar-nav me-auto mb-2 mb-lg-0"
            >
              <li
                class="nav-item"
              >
                <a
                  aria-current="page"
                  class="nav-link active"
                  data-discover="true"
                  href="/"
                >
                  Home
                </a>
              </li>
              <li
                class="nav-item"
              >
                <a
                  class="nav-link"
                  data-discover="true"
                  href="/leaderboard"
                >
                  Leaderboard
                </a>
              </li>
              <li
                class="nav-item"
              >
                <a
                  class="nav-link"
                  data-discover="true"
                  href="/add"
                >
                  New
                </a>
              </li>
            </ul>
            <span
              class="d-flex"
            >
              <span
                class="d-flex align-items-center justify-content-center"
              >
                <img
                  height="30"
                  src="/default-user.jpg"
                  width="30"
                />
              </span>
              <button
                class="btn"
              >
                Log out
              </button>
            </span>
          </div>
        </div>
      </nav>
      <main />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
